src/Microsoft.AspNet.Mvc.Core/ControllerContext.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNet.Mvc.Controllers;
+using Microsoft.AspNet.Mvc.Core;
+using Microsoft.AspNet.Mvc.Formatters;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Microsoft.AspNet.Mvc.ModelBinding.Validation;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// The context associated with the current request for a controller.
+    /// </summary>
+    public class ControllerContext : ActionContext
+    {
+        private FormatterCollection<IInputFormatter> _inputFormatters;
+        private IList<IModelBinder> _modelBinders;
+        private IList<IModelValidatorProvider> _validatorProviders;
+        private IList<IValueProvider> _valueProviders;
+
+        /// <summary>
+        /// Creates a new <see cref="ControllerContext"/>. This constructor should only be used for
+        /// unit testing.
+        /// </summary>
+        public ControllerContext()
+            : base()
+        {
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="ControllerContext"/>.
+        /// </summary>
+        /// <param name="context">The <see cref="ActionContext"/> associated with the current request.</param>
+        public ControllerContext(ActionContext context)
+            : base(context)
+        {
+            if (!(context.ActionDescriptor is ControllerActionDescriptor))
+            {
+                throw new ArgumentException(Resources.FormatActionDescriptorMustBeBasedOnControllerAction(
+                    typeof(ControllerActionDescriptor)),
+                    nameof(context));
+            }
+        }
+
+        public new ControllerActionDescriptor ActionDescriptor
+        {
+            get { return (ControllerActionDescriptor)base.ActionDescriptor; }
```

Comment:
Should this account for `ActionContext.ActionDescriptor` being independently settable?